### PR TITLE
refactor(xmr): remove old BP code

### PR DESCRIPTION
**Pull request by ph4r05**
- Removes old Bulletproof code and constants

### CODE REVIEW

**Comment by grdddj** [comment]
The results look really nice!

It seems to save over `20 kb` of our precious flash space.

Full diff of firmware binary before and after this commit:
```
.flash2    17      3    L1       D16      src/apps/monero/signing/step_06_set_output.py                _rsig_bp()
.flash2    15      3    L-1      D16      src/apps/monero/signing/step_01_init_transaction.py          _check_rsig_data()
.flash2    3       3    L-13     D16      src/apps/monero/signing/step_06_set_output.py                _rsig_process_bp()
.flash2    -2      1    L-2      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder
.flash2    -2      1    L-2      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder.__init__()
.flash2    -2      1    L-2      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder.gc()
.flash2    -2      1    L-2      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder.prove()
.flash2    -2      1    L-2      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder.prove_batch()
.flash2    -2      1    L-2      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder.vector_exponent()
.flash2    -2      1    L-2      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder.verify()
.flash2    -2      1    L-2      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusData
.flash2    -2      1    L-2      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusData.__init__()
.flash2    -2      1    L-2      D0       src/apps/monero/xmr/bulletproof.py                           _compute_LR()
.flash2    -4      2    L-4      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder._gprec_aux()
.flash2    -4      2    L-4      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder._hprec_aux()
.flash2    -4      2    L-4      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder.prove_setup_aLaR()
.flash2    -4      2    L-4      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder.verify_batch()
.flash2    -4      1    L-4      D0       src/apps/monero/xmr/range_signatures.py                      
.flash2    -5      1    L-5      D0       src/apps/monero/signing/__init__.py                          RctType
.flash2    -6      3    L-6      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder.prove_setup()
.flash     -8      1    L-8      D0       mp_qstr_const_pool
.flash2    -8      4    L-8      D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder.aX_vcts()
.flash2    -12     6    L-12     D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder._prove_batch_main()
.flash     -16     1    L-16     D0       mod_trezorcrypto_monero_BP_GI_PRE_obj
.flash     -16     1    L-16     D0       mod_trezorcrypto_monero_BP_HI_PRE_obj
.flash     -16     1    L-16     D0       mod_trezorcrypto_monero_globals_table
.flash2    -18     2    L-18     D0       src/apps/monero/xmr/range_signatures.py                      prove_range_bp_batch()
.flash2    -21     1    L-21     D0       src/apps/monero/xmr/range_signatures.py                      verify_bp()
.flash2    -118    1    L-118    D0       src/apps/monero/xmr/mlsag_hasher.py                          PreMlsagHasher.rsig_val()
.flash2    -188    1    L-188    D0       mp_qstr_frozen_const_pool
.flash2    -239    8    L-239    D0       src/apps/monero/xmr/serialize_messages/tx_rsig_bulletproof.py 
.flash2    -285    5    L-285    D0       src/apps/monero/signing/step_06_set_output.py                
.flash2    -935    1    L-935    D0       [section .flash2]
.flash     -8_136  1    L-8_136  D0       [section .flash]
.flash2    -10_960 200  L-10_632 D-328    src/apps/monero/xmr/bulletproof.py
SUMMARY: 35 rows, -20_992 bytes in total (L-20_712 D-280).
```

I have re-pushed the branch into `firmware` repo, so we can see the `CI` - https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/669994787

It looks like even more code removing is possible, I have run the `Vulture` tool over `src/apps/monero` and it found many (probable) dead code:
```
src/apps/monero/signing/offloading_keys.py:84: unused function 'hmac_key_txout_asig'
src/apps/monero/signing/offloading_keys.py:105: unused function 'enc_key_cout'
src/apps/monero/signing/offloading_keys.py:119: unused function 'det_comm_masks'
src/apps/monero/xmr/addresses.py:32: unused function 'decode_addr'
src/apps/monero/xmr/addresses.py:42: unused function 'public_addr_encode'
src/apps/monero/xmr/bulletproof.py:143: unused function '_scalarmultH'
src/apps/monero/xmr/bulletproof.py:203: unused function '_sc_mul8'
src/apps/monero/xmr/bulletproof.py:238: unused function '_sc_mulsub'
src/apps/monero/xmr/bulletproof.py:257: unused function '_sub_keys'
src/apps/monero/xmr/bulletproof.py:451: unused method 'slice'
src/apps/monero/xmr/bulletproof.py:630: unused method 'realloc_init_from'
src/apps/monero/xmr/bulletproof.py:694: unused class 'KeyVSized'
src/apps/monero/xmr/bulletproof.py:713: unused class 'KeyVConst'
src/apps/monero/xmr/bulletproof.py:1201: unused function '_cross_inner_product'
src/apps/monero/xmr/bulletproof.py:1284: unused method 'add_scalar'
src/apps/monero/xmr/bulletproof.py:1418: unused method 'vector_exponent'
src/apps/monero/xmr/crypto_helpers.py:122: unused function 'generate_signature'
src/apps/monero/xmr/crypto_helpers.py:140: unused function 'check_signature'
src/apps/monero/xmr/serialize/__init__.py:33: unused function 'dump_msg_gc'
src/apps/monero/xmr/serialize/base_types.py:40: unused class 'XmrStructuredType'
src/apps/monero/xmr/serialize/base_types.py:67: unused class 'UInt8'
src/apps/monero/xmr/serialize/int_serialize.py:45: unused function 'load_uvarint_b'
src/apps/monero/xmr/serialize/int_serialize.py:82: unused function 'dump_uint_b_into'
src/apps/monero/xmr/serialize/message_types.py:20: unused class 'UnicodeType'
src/apps/monero/xmr/serialize/readwriter.py:32: unused method 'is_empty'
```

I did these deletion in https://github.com/trezor/trezor-firmware/commit/1a451b7e5d7db953f1491fffc4ce025708df7bc8 and they seem to save `3.5 kb` with this binary diff:
```
.flash2    -16     2    L-16     D0       src/apps/monero/xmr/bulletproof.py                           KeyV
.flash2    -16     2    L-16     D0       src/apps/monero/xmr/serialize/__init__.py                    
.flash2    -26     2    L-26     D0       src/apps/monero/xmr/serialize/int_serialize.py               
.flash2    -59     5    L-59     D0       src/apps/monero/xmr/bulletproof.py                           MultiExpSequential
.flash2    -59     4    L-59     D0       src/apps/monero/xmr/serialize/base_types.py                  
.flash2    -60     4    L-44     D-16     src/apps/monero/signing/offloading_keys.py                   hmac_key_txout()
.flash2    -60     3    L-60     D0       src/apps/monero/xmr/serialize/__init__.py                    dump_msg()
.flash2    -62     5    L-62     D0       src/apps/monero/xmr/serialize/readwriter.py                  MemoryReaderWriter
.flash2    -74     3    L-74     D0       src/apps/monero/xmr/serialize/int_serialize.py               load_uvarint()
.flash2    -75     3    L-75     D0       src/apps/monero/xmr/serialize/int_serialize.py               dump_uint()
.flash2    -78     5    L-78     D0       src/apps/monero/xmr/bulletproof.py                           KeyVBase
.flash2    -82     5    L-82     D0       src/apps/monero/xmr/bulletproof.py                           BulletProofPlusBuilder
.flash2    -124    1    L-124    D0       mp_qstr_frozen_const_pool
.flash2    -162    8    L-162    D0       src/apps/monero/xmr/addresses.py                             
.flash2    -166    10   L-134    D-32     src/apps/monero/signing/offloading_keys.py                   
.flash2    -199    11   L-199    D0       src/apps/monero/xmr/serialize/message_types.py               
.flash2    -309    4    L-293    D-16     src/apps/monero/xmr/bulletproof.py                           KeyV.realloc()
.flash2    -329    9    L-313    D-16     src/apps/monero/xmr/crypto_helpers.py                        
.flash2    -370    1    L-370    D0       [section .flash2]
.flash2    -1_258  41   L-1_258  D0       src/apps/monero/xmr/bulletproof.py                           
SUMMARY: 20 rows, -3_584 bytes in total (L-3_504 D-80).
```

**Review by prusnak — APPROVED** [review]
Changes in `core/embed/extmod/modtrezorcrypto` ACK

**Review by matejcik — COMMENTED** [review]
thanks! nice, clean, easy to review :+1: :)

File: core/src/apps/monero/signing/step_01_init_transaction.py [hunk 1]
```diff
@@ -207,10 +207,11 @@ def _check_rsig_data(state: State, rsig_data: MoneroTransactionRsigData) -> None
...
 
-    state.tx_type = signing.RctType.CLSAG
-    if rsig_data.bp_version == 4:
-        state.rsig_is_bp_plus = True
-        state.tx_type = signing.RctType.RCTTypeBulletproofPlus
+    state.tx_type = signing.RctType.RCTTypeBulletproofPlus
+    if rsig_data.bp_version != 4:
+        raise ValueError("Unknown BP version")
+
+    state.rsig_is_bp_plus = True
```
[the remark this answers is not in the document]
Comment (ph4r05) [reply core/src/apps/monero/signing/step_01_init_transaction.py:214]: Thats a good point! 😅  I will add fixup for this (sqash or commit and squash later?) removing also functions indicated by @grdddj (nice report!)

[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/monero/signing/step_01_init_transaction.py:214]: if you're doing more changes, please make a separate fixup and squash after the review is approved

(that or we'll do squash-merge)

**Comment by ph4r05** [comment]
I've cherry-picked @grdddj commits, added the one removing [rsig_is_bp_plus](https://github.com/trezor/trezor-firmware/pull/2570/commits/dfc319b7916fbe98df791605ffd3d3420728432c). 

Both Monero unit tests and Monero Wallet pass

**Comment by grdddj** [comment]
@ph4r05 Thanks for the `flake` changes, I was about to commit them as well.

There is also a need to run `make gen`, because of deleted `BP_GI_PRE` and `BP_HI_PRE` attributes, which are autogenerated into a `.pyi` file. CI failure - https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/3190875570

**Comment by grdddj** [comment]
> ad [monero.pyi](https://github.com/trezor/trezor-firmware/pull/2570/files#diff-1d1182a6eecdcc6897bad258f125fa0749d80f90385645ae9e010ef64b7762a2), I think I already fixed this in this PR, isn't CI result from an older version?

You are right, my mistake :) I will push your branch into our repo, so we run `Gitlab` on it in the current state. Will there be some more changes?

**Comment by ph4r05** [comment]
@grdddj thanks! I believe this will be all.

**Comment by mmilata** [comment]
Squashed & merged. Thanks @ph4r05!

File: core/tests/run_tests_device_emu_monero.sh [hunk 2]
```diff
@@ -50,7 +50,7 @@ elif [[ "$OSTYPE" == "darwin"* || "$FORCE_DOCKER_USE" == 1 ]]; then
   docker exec $DOCKER_ID apt-get install --no-install-recommends --no-upgrade -qq net-tools socat 2>/dev/null >/dev/null
   docker exec -d $DOCKER_ID socat UDP-LISTEN:21324,reuseaddr,reuseport,fork UDP4-SENDTO:host.docker.internal:21324
   docker exec -d $DOCKER_ID socat UDP-LISTEN:21325,reuseaddr,reuseport,fork UDP4-SENDTO:host.docker.internal:21325
-  docker exec $DOCKER_ID "$TREZOR_MONERO_TESTS_PATH" 2>&1 > "$TREZOR_MONERO_TESTS_LOG"
+  docker exec -e TEST_MAX_HF=15 -e TEST_MIN_HF=15 $DOCKER_ID - "$TREZOR_MONERO_TESTS_PATH" 2>&1 > "$TREZOR_MONERO_TESTS_LOG"
```
Comment (ph4r05) [inline core/tests/run_tests_device_emu_monero.sh:53]: Hmm @mmilata pls check this out, the `-` in `$DOCKER_ID -` may be a typo.

Comment (mmilata) [reply core/tests/run_tests_device_emu_monero.sh:53]: #2584

**Pull request merged**